test/task_graph_tests.cc

```diff
@@ -0,0 +1,448 @@
+		auto buf_a = mbf.create_buffer(cl::sycl::range<1>(128));
+
+		SECTION("with true dependencies") {
+			test_utils::add_host_task(tm, on_master_node, [&](handler& cgh) { buf_a.get_access<mode::discard_write>(cgh, fixed<1>({0, 128})); });
+			test_utils::add_host_task(tm, on_master_node, [&](handler& cgh) { buf_a.get_access<mode::read_write>(cgh, fixed<1>({0, 128})); });
+
+			auto* previous_horizon = task_manager_testspy::get_previous_horizon_task(tm);
+			CHECK(previous_horizon == nullptr);
+
+			const auto tid_c = test_utils::add_host_task(tm, on_master_node, [&](handler& cgh) { buf_a.get_access<mode::read>(cgh, fixed<1>({0, 128})); });
+
+			previous_horizon = task_manager_testspy::get_previous_horizon_task(tm);
+			REQUIRE(previous_horizon != nullptr);
+			CHECK(previous_horizon->get_id() == tid_c + 1);
+			CHECK(task_manager_testspy::get_num_horizons(tm) == 1);
+
+			auto horizon_dependencies = previous_horizon->get_dependencies();
+
+			CHECK(std::distance(horizon_dependencies.begin(), horizon_dependencies.end()) == 1);
+			CHECK(horizon_dependencies.begin()->node->get_id() == tid_c);
+
+			std::set<task_id> expected_dependency_ids;
+
+			// previous horizon is always part of the active task front
+			expected_dependency_ids.insert(previous_horizon->get_id());
+			expected_dependency_ids.insert(test_utils::add_host_task(tm, on_master_node, [&](handler& cgh) {}));
+			expected_dependency_ids.insert(test_utils::add_host_task(tm, on_master_node, [&](handler& cgh) {}));
+			expected_dependency_ids.insert(test_utils::add_host_task(tm, on_master_node, [&](handler& cgh) {}));
+			CHECK(task_manager_testspy::get_num_horizons(tm) == 1);
+
+			test_utils::add_host_task(tm, on_master_node, [&](handler& cgh) { buf_a.get_access<mode::read_write>(cgh, fixed<1>({0, 128})); });
+			const auto tid_d = test_utils::add_host_task(tm, on_master_node, [&](handler& cgh) {
+				buf_a.get_access<mode::read_write>(cgh, fixed<1>({0, 128}));
+			});
+			expected_dependency_ids.insert(tid_d);
+
+			previous_horizon = task_manager_testspy::get_previous_horizon_task(tm);
+			REQUIRE(previous_horizon != nullptr);
+			CHECK(previous_horizon->get_id() == tid_d + 1);
+			CHECK(task_manager_testspy::get_num_horizons(tm) == 2);
+
+			horizon_dependencies = previous_horizon->get_dependencies();
+			CHECK(std::distance(horizon_dependencies.begin(), horizon_dependencies.end()) == 5);
+
+			std::set<task_id> actual_dependecy_ids;
+			for(auto dep : horizon_dependencies) {
+				actual_dependecy_ids.insert(dep.node->get_id());
+			}
+			CHECK(expected_dependency_ids == actual_dependecy_ids);
+		}
+	}
+
+	static inline GridRegion<3> make_region(int min, int max) { return GridRegion<3>(GridPoint<3>(min, 0, 0), GridPoint<3>(max, 1, 1)); }
+
+	TEST_CASE("task horizons update previous writer data structure", "[task_manager][task-graph][task-horizon]") {
+		task_manager tm{1, nullptr, nullptr};
+		tm.set_horizon_step(2);
+
+		using namespace cl::sycl::access;
```

Comment:
Since we have `celerity::access_mode` now, we should prefer it over `cl::sycl::access::mode`!